x/oracle/keeper/keeper.go

```diff
@@ -179,7 +130,26 @@ func (k Keeper) GetExchangeRateTwap(ctx sdk.Context, pair asset.Pair) (price sdk
 		return sdk.OneDec().Neg(), types.ErrNoValidTWAP
 	}
 
-	return k.calcTwap(ctx, snapshots)
+	if len(snapshots) == 1 {
+		return snapshots[0].Price, nil
+	}
+
+	firstTimestampMs := snapshots[0].TimestampMs
+
+	cumulativePrice := sdk.ZeroDec()
+	for i, s := range snapshots {
+		var nextTimestampMs int64
+		if i == len(snapshots)-1 {
+			// if we're at the last snapshot, then consider that price as ongoing until the current blocktime
+			nextTimestampMs = ctx.BlockTime().UnixMilli()
+		} else {
+			nextTimestampMs = snapshots[i+1].TimestampMs
+		}
+
+		price := s.Price.MulInt64(nextTimestampMs - s.TimestampMs)
+		cumulativePrice = cumulativePrice.Add(price)
+	}
+	return cumulativePrice.QuoInt64(ctx.BlockTime().UnixMilli() - firstTimestampMs), nil
```

Comment:
Please add an explicit error that rules out zero division if the `snapshots[0].TimestampMs` is equal to the current block time

Reply:
Fixed.